### ISSUE TITLE
Issues w/ softmax procedure

### ISSUE THREAD

**Issue by Niminem** [mratsim/Arraymancer#554]
Firstly, I had to modify the arraymancer import to exclude 'softmax' procedure and then include it separately like so:
`import arraymancer except softmax`
`import arraymancer/nn/activation/softmax`

If I don't do this, the only version I'm getting is this one:
`nnp_softmax.softmax: proc (input: Tensor[softmax.T]): Tensor[softmax.T]`

Which would throw an error because it accepts `Tensor`s where as I need the version that accepts `Variable[TT]`, defined here: https://github.com/mratsim/Arraymancer/blob/master/src/arraymancer/nn/activation/softmax.nim
(because this is needed in my custom `forward` procedure)

When I did this, I'm STILL getting an error thrown although it's something different:

/Users/salient/Desktop/CatsVDogs/main.nim(75, 26) template/generic instantiation of `forward` from here
/Users/salient/Desktop/CatsVDogs/main.nim(63, 9) template/generic instantiation of `softmax` from here
/Users/salient/.nimble/pkgs/arraymancer-0.7.11/arraymancer/nn/activation/softmax.nim(58, 11) template/generic instantiation of `softmax_cache` from here
/Users/salient/.nimble/pkgs/arraymancer-0.7.11/arraymancer/nn/activation/softmax.nim(42, 24) template/generic instantiation of `softmax_backward_ag` from here
/Users/salient/.nimble/pkgs/arraymancer-0.7.11/arraymancer/nn/activation/softmax.nim(23, 35) Error: type mismatch: got <type Gate> but expected 'SoftmaxActivation[Tensor[system.float32]]'

Any explanation for what this is and/or how to fix it?

I do see an open issue that looks somewhat relevant but not too sure: https://github.com/mratsim/Arraymancer/issues/472

**Comment by Vindaar** [comment]
Could you provide a small reproducible example?

**Comment by Niminem** [comment]
import std/[random, os]
import arraymancer except softmax # having issue w/ softmax, see below for ref
import arraymancer/nn/activation/softmax

randomize(9000) # seed the random number generator for repeatable results

const dir = "training_data" # directory where training data is stored (relative to this file)
const BatchSize = 32 # number of images to process in a batch

let ctx = newContext Tensor[float32] # autograd/neural network graph

var X = ctx.variable read_npy[float32](dir / "X.npy").unsqueeze(1) # shape [24946, 1, 50, 50]
# orignially of shape [24946, 50, 50]
# changed shape from [N, H, W] to [N, C, H, W], with C = 1 (unsqueeze). Convolution expect 4d tensors
# var y = read_npy[int](dir / "Y.npy") # shape # [24946, 2]

type
    Kernal = object
        weight, bias: Variable[Tensor[float32]] # remove bias if found unnecessary
    Model = object
        k1, k2, k3: Kernal
        hidden, classifier: LinearLayer[Tensor[float32]]

template weightInit(shape: varargs[int], init_kind: untyped): Variable =
    ctx.variable(init_kind(shape, float32), requires_grad = true)

proc newModel(ctx: Context[Tensor[float32]]): Model =
    result.k1.weight = weightInit(32, 1, 5, 5, kaiming_normal)  # conv2d kernal shape [C_out,C_in,kH,kW]
    result.k1.bias = ctx.variable(zeros[float32](32, 1, 1), requires_grad = true) # "Nilable" bias shape [C_out,1,1]

    result.k2.weight = weightInit(64, 32, 5, 5, kaiming_normal)
    result.k2.bias = ctx.variable(zeros[float32](64, 1, 1), requires_grad = true)

    result.k3.weight = weightInit(128, 64, 5, 5, kaiming_normal)
    result.k3.bias = ctx.variable(zeros[float32](128, 1, 1), requires_grad = true)

    result.hidden.weight = weightInit(512, 512, kaiming_normal)
    result.hidden.bias = ctx.variable(zeros[float32](1, 512), requires_grad = true)

    result.classifier.weight = weightInit(2, 512, yann_normal)
    result.classifier.bias = ctx.variable(zeros[float32](1, 2), requires_grad = true)

proc forward(m: Model, x: Variable): Variable = # is var Model parameter better here?
    result = x
        # bias necessary for conv2d? it is optional in arraymancer
        .conv2d(m.k1.weight, m.k1.bias, padding=(0,0), stride=(1,1)) # padding & stride are default values (shown for clarity)
        .relu
        .maxpool2d(kernel= (2,2), padding= (0,0), stride= (2,2)) # is stride necessary?
        .conv2d(m.k2.weight, m.k2.bias)
        .relu
        .maxpool2d(kernel= (2,2), padding= (0,0), stride= (2,2))
        .conv2d(m.k3.weight, m.k3.bias)
        .relu
        .maxpool2d(kernel= (2,2), padding= (0,0), stride= (2,2))
        .flatten # shape is [32, 512], is this the formula --> [N, C * H * W] ?
        .linear(m.hidden.weight, m.hidden.bias)
        .relu
        .linear(m.classifier.weight, m.classifier.bias)
        .softmax
        # *** potential issue w/ softmax: https://github.com/mratsim/Arraymancer/issues/472

let model = ctx.newModel
# Learning loop
for epoch in 0 ..< 5:
    for batch_id in 0 ..< X.value.shape[0] div BatchSize: # some at the end may be missing, oh well ...
        # minibatch offset in the Tensor
        let offset = batch_id * BatchSize
        let x = X[offset ..< offset + BatchSize, _]

        # Running through the network and computing loss
        let yPred = model.forward(x)
        echo yPred.value.shape
        quit() # **** remove me ****

**Comment by Niminem** [comment]
@Vindaar the above is the "simple 2 layer" example modified to simply add softmax in the forward. It produces the same error as above

**Comment by Vindaar** [comment]
Thanks! I felt so free as to update your comment and turn it into an actual code snippet.

Will check it out.

**Comment by Niminem** [comment]
Thanks Vindaar for your fast response and comment edit lol I'm still getting used to Github markdown

**Comment by Niminem** [comment]
I've modified the `softmax_backward_ag[TT]` procedure to pass in `self` rather than `Gate`: (see below)
reference: https://github.com/mratsim/Arraymancer/blob/master/src/arraymancer/nn/activation/softmax.nim
```nim
proc softmax_backward_ag[TT](self: Gate[TT], payload: Payload[TT]): SmallDiffs[TT] =
  let self = SoftmaxActivation[TT](self)#(Gate)
  let gradient = payload.variable.grad
  result = newDiffs[TT](1)
  result[0] = gradient.softmax_backward(self.cache)
```

This matches what I've found while looking at how `relu` is implemented. It took care of this error that was raised:
`type mismatch: got <type Gate> but expected 'SoftmaxActivation[Tensor[system.float32]]'`

However, now I get this error:
`attempting to call undeclared routine: 'softmax_backward'`

After a search through the docs I see we don't have the `softmax_backward` procedure as mentioned in this issue:
https://github.com/mratsim/Arraymancer/issues/472

**Comment by Niminem** [comment]
@Vindaar please review when you can

**Comment by Niminem** [comment]
@Vindaar my good sir can we please get this implemented lol

**Comment by Vindaar** [comment]
Can you please ping me about this on matrix/discord on the weekend, if I haven't looked into this by then?

**Comment by Vindaar** [comment]
Ok, I just had a look at it. 

As you've mentioned yourself, the practical problem is that the backward pass for `softmax` is not implemented. After looking into it now, I realize that the (likely) reason for that is that the backward pass of a pure `softmax` is rather ugly, because the softmax itself is defined via a sum of all parameters. In essence then the gradient results in something that depends on the specific indices (you have a `δij` in the derivative):
```
∂sm(x_i) / ∂x_j = sm(x_j) · ( δij - sm(x_i) )
```
(sorry for somewhat sloppy notation)

See for example:
https://en.wikipedia.org/wiki/Softmax_function

That's why typically one combines the softmax on the last layer directly with a cross entropy loss, for which the gradient is easy to compute. 

I don't have the time & mental space atm to figure out how to efficiently implement this (if even possible?). If someone is willing to do so, feel free. Otherwise I'd just recommend to do what one normally does, i.e. use `softmax_cross_entropy`)

**Comment by Niminem** [comment]
Shit, thanks for looking into it Vindaar. I will take a look when I finally get the time and mental space as well lol
